ttnn/cpp/ttnn/tensor/layout/tensor_layout.cpp

```diff
@@ -122,33 +159,104 @@ size_t TensorLayout::compute_page_size_bytes(const Size& page_size) const {
     const int rank = static_cast<int>(shape.rank());
     const int alignment_rank = static_cast<int>(alignment_.size());
-    const int max_rank = std::max(rank, alignment_rank);
+
     size_t width = 1;
     size_t height = 1;
 
-    // Iterate dims in reverse order and ensure alignment
-    // Even tensor of rank 0 or 1 must be aligned (to Tile / Page / Shard)
-    for (int i = -1; i >= -max_rank; --i) {
-        auto& dim = i == -1 ? width : height;
-        if(i >= -rank) {
+    if (memory_config_.shard_spec.has_value() and memory_config_.shard_spec.value().mode == ShardMode::LOGICAL) {
+        // Iterate dims in reverse order
+        for (int i = -1; i >= -rank; --i) {
+            auto& dim = i == -1 ? width : height;
             dim *= shape[i];
         }
 
-        // Align the current dimension if alignment is available
-        if (i >= -alignment_rank) {
-            dim = CMAKE_UNIQUE_NAMESPACE::round_up(dim, alignment_[i]);
+        auto compute_default_logical_shard_shape = [&shape, &rank]() {
+            size_t shard_width = 1;
+            size_t shard_height = 1;
+
+            if (rank >= 1) {
+                shard_width = shape[-1];
+            }
+            if (rank >= 2) {
+                shard_height = shape[-2];
+            }
+
+            return Size{shard_height, shard_width};
+        };
+
+        const auto& logical_shard_shape = memory_config_.shard_spec.has_value() ? Size(memory_config_.shard_spec.value().shape) : compute_default_logical_shard_shape();
+        const auto& physical_shard_shape = compute_physical_shard_shape(logical_shard_shape);
+
+        auto get_physical_size = [](auto original_size, auto logical_shard_size, auto physical_shard_size, auto alignment) -> uint32_t {
+            if (logical_shard_size == 0) {
+                return 0;
+            }
+            // If we always pad to full shards, then return:
+            // auto num_shards = tt::div_up(original_size, logical_shard_size);
+            // return (uint32_t) physical_shard_size * num_shards;
+
+            // But host physical data is only padded up to nearest alignment (and no padding in between shards)
+            // So last shard is only up to nearest alignment
+            auto num_full_shards = original_size / logical_shard_size;
+            auto last_physical_shard_size = CMAKE_UNIQUE_NAMESPACE::round_up(original_size % logical_shard_size, alignment);
+            return (physical_shard_size * num_full_shards + last_physical_shard_size);
+        };
+
+        auto physical_height = get_physical_size(height, logical_shard_shape.height(), physical_shard_shape.height(), alignment_[-2]);
+        auto physical_width = get_physical_size(width, logical_shard_shape.width(), physical_shard_shape.width(), alignment_[-1]);
+
+        Size size{physical_height, physical_width};
+        return size;
+    } else {
+        const int max_rank = std::max(rank, alignment_rank);
```

Comment:
Can you please add a comment above this line saying which case are we handling?
For example: `handling non-sharded layout`